### PR TITLE
Issue 83: Reorganize repository structure

### LINKED ISSUE #83: Reorganize repository structure
**Problem description**
Currently, the repository lacks from a clear structure that distinguishes between components (standalone, flink, hadoop), features (state synchronizer) and scenarios (IoT, anomaly detection).

**Problem location**
Repository structure.

**Suggestions for an improvement**
Reorganize the repository structure in a way that can be cleanly extended with potentially diverse types of examples.

### PR DESCRIPTION
**Pull request by RaulGracia**
**Change log description**
This PR represents a major change in how sample applications are organized. A general overview of the new structure can be seen in this [wiki page](https://github.com/pravega/pravega-samples/wiki/Repository-Structure-Reorganization). In summary, existing samples have been organized as follows:

    pravega-client-examples
        - gettingstarted
        - consolerw
        - noop
        - statesynchronizer
        - streamcuts

    flink-connector-examples
        - wordcount

    hadoop-connector-examples
        - wordcount

    scenarios
        - turbine-heat-sensor	
        - turbine-heat-processor
        - anomaly-detection

The idea is to have simple and small examples within `xxx-examples` projects based on the type of example at hand, whereas more complex, potentially polyglot examples are organized as separate projects in the `scenarios` folder. This change also implied to modify accordingly the `build` and `readme` files. 

**Purpose of the change**
Fixes #83.

**What the code does**
Change the structure of samples repository.

**How to verify it**
Applications should run as before with the new repository structure.

### CODE REVIEW

**Comment by RaulGracia** [comment]
[Reorganize repository structure (pravega-samples)](https://trello.com/c/1XrWLS53/43-reorganize-repository-structure-pravega-samples)

File: flink-connector-examples/src/main/dist/bin/run-example [hunk 1]
```diff
@@ -0,0 +1,3 @@
+#!/usr/bin/env bash
+
+flink run -c io.pravega.example.flink.turbineheatprocessor.TurbineHeatProcessor lib/pravega-flink-examples-0.1.0-SNAPSHOT-all.jar $*
```
Comment (Prabha-Veerubhotla) [inline flink-connector-examples/src/main/dist/bin/run-example]: add new line here.

File: gradle.properties [hunk 2]
```diff
@@ -13,10 +13,16 @@ dcosAddress=master.mesos
...
 samplesVersion=0.3.0-SNAPSHOT
 
-### Flink
+### Flink-connector examples
 flinkVersion=1.4.0
+
+### Hadoop-connector examples
+hadoopVersion=2.8.1
+scalaVersion=2.11.8
+sparkVersion=2.2.0
```
Comment (Prabha-Veerubhotla) [inline gradle.properties]: new line

File: pravega-client-examples/README.md [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+It is worthwhile to launch two or more separate CLIs in separate windows using the same settings and observe how changes in one
+CLI process are not visible in another CLI process until that other CLI process invokes REFRESH.
+
+## NoopReader
+
+An example of a simple reader that continually reads the contents of any stream. A binary serializer is used so it works against any event types. The sample emits basic information about number of events/bytes read every 30 seconds. 
+
+```
+$ bin/noopReader [--uri tcp://127.0.0.1:9090] [--stream <SCOPE>/<STREAM>]
+```
```
Comment (Prabha-Veerubhotla) [inline pravega-client-examples/README.md]: Can we include stream cuts also? if there is way to execute them.
Also, it will be great if we can follow the order mentioned in the https://github.com/pravega/pravega-samples/wiki/Repository-Structure-Reorganization#repository-structure-proposal

File: scenarios/turbine-heat-processor/src/main/dist/bin/run-example [hunk 4]
```diff
@@ -0,0 +1,3 @@
+#!/usr/bin/env bash
+
+flink run -c io.pravega.example.flink.turbineheatprocessor.TurbineHeatProcessor lib/pravega-flink-examples-0.1.0-SNAPSHOT-all.jar $*
```
Comment (Prabha-Veerubhotla) [inline scenarios/turbine-heat-processor/src/main/dist/bin/run-example]: new line

File: scenarios/turbine-heat-sensor/src/main/java/io/pravega/turbineheatsensor/TurbineHeatSensor.java [hunk 5]
```diff
@@ -457,4 +457,4 @@ public void run() {
             return clientFactory.createReader(readerName, readerGroup, SERIALIZER, readerConfig);
         }
     }
-}
+}
```
Comment (Prabha-Veerubhotla) [inline scenarios/turbine-heat-sensor/src/main/java/io/pravega/turbineheatsensor/TurbineHeatSensor.java]: new line

File: README.md [hunk 6]
```diff
@@ -11,56 +11,66 @@ developer guide](http://pravega.io).
 
 # Repository Structure
 
-This repository is divided into sub-projects (`standalone-examples`, `flink-examples`
-and `hadoop-examples`) addressed to demonstrate a specific component. In these sub-projects, 
+This repository is divided into sub-projects (`pravega-client-examples`, `flink-connector-examples`
+and `hadoop-connector-examples`) addressed to demonstrate a specific component. In these sub-projects,
```
[the remark this answers is not in the document]
Comment (RaulGracia) [reply README.md]: Scenarios contains sub-project, but it actually is a folder. I changed that in the description.

File: README.md [hunk 7]
```diff
@@ -11,56 +11,66 @@ developer guide](http://pravega.io).
...
-and `hadoop-examples`) addressed to demonstrate a specific component. In these sub-projects, 
+This repository is divided into sub-projects (`pravega-client-examples`, `flink-connector-examples`
+and `hadoop-connector-examples`) addressed to demonstrate a specific component. In these sub-projects, 
 we provide a battery of simple code examples aimed at demonstrating how a particular 
 feature or API works. Moreover, we also include a `scenarios` sub-project that contains 
 more complex applications, which show use-cases exploiting one or multiple components.
 
+> Hint: Have a look to the [terminology and concepts](http://pravega.io/docs/latest/terminology/) in Pravega.
+
 ## Pravega Examples
```
[the remark this answers is not in the document]
Comment (RaulGracia) [reply README.md]: Done.

File: README.md [hunk 8]
```diff
@@ -71,14 +81,15 @@ $ ./gradlew install
 ```
 
 The above command should generate the required `jar` files into your local Maven repository.
-For more information, please visit [Pravega](https://github.com/pravega/pravega). 
 
 > Hint: If you use a different version of Pravega, please check the `pravegaVersion` property
```
[the remark this answers is not in the document]
Comment (RaulGracia) [reply README.md]: Done.

File: README.md [hunk 9]
```diff
@@ -87,11 +98,14 @@ $ cd flink-connectors
 $ ./gradlew clean install
 ```
 
+> Hint: If you use a different version of Flink connector, please check the `flinkConnectorVersion` property
```
[the remark this answers is not in the document]
Comment (RaulGracia) [reply README.md]: Done.

File: README.md [hunk 10]
```diff
@@ -100,6 +114,9 @@ $ cd hadoop-connectors
 $ gradle install
 ```
 
+> Hint: If you use a different version of Hadoop connector, please check the `hadoopConnectorVersion` property
```
[the remark this answers is not in the document]
Comment (RaulGracia) [reply README.md]: Done.

File: README.md [hunk 11]
```diff
@@ -118,24 +135,24 @@ scripts. The above gradle command automatically creates the execution scripts th
...
+pravega-samples/pravega-client-examples/build/install/pravega-client-examples/bin
 ```
 
 There is a Linux/Mac script and a Windows (.bat) script for each separate executable.
 
 # Proposed Roadmap
 
 We propose a roadmap to proceed with the execution of examples based on their complexity:
-1. [Pravega examples](https://github.com/pravega/pravega-samples/tree/master/standalone-examples): 
+1. [Pravega examples](https://github.com/pravega/pravega-samples/tree/master/pravega-client-examples):
```
[the remark this answers is not in the document]
Comment (RaulGracia) [reply README.md]: Done.

File: flink-connector-examples/README.md [hunk 12]
```diff
@@ -0,0 +1,144 @@
...
+The `upload` task makes it easy to upload the sample binaries to your cluster. First, configure 
+Gradle with the address of a node in your cluster.   Edit `~/.gradle/gradle.properties` to 
+specify a value for `dcosAddress`.
+
+```
+$ cat ~/.gradle/gradle.properties
+dcosAddress=10.240.124.164
+```
+
+Then, upload the samples to the cluster.  They'll be copied to `/home/centos` on the target node.
```
[the remark this answers is not in the document]
Comment (RaulGracia) [reply flink-connector-examples/README.md]: Done.

File: flink-connector-examples/README.md [hunk 13]
```diff
@@ -0,0 +1,144 @@
...
+
+---
+
+# Examples Catalog
+
+## Word Count
+
+This example consists of two applications, a `WordCountWriter` that reads data from a 
+network stream, transforms the data, and writes the data to a Pravega stream; and a
+`WordCountReader` that reads from a Pravega stream and prints the word counts summary.
```
[the remark this answers is not in the document]
Comment (RaulGracia) [reply flink-connector-examples/README.md]: Done.

File: flink-connector-examples/src/main/dist/bin/run-example (hunk 1 above)
Comment (RaulGracia) [reply flink-connector-examples/src/main/dist/bin/run-example]: Added.

File: gradle.properties (hunk 2 above)
Comment (RaulGracia) [reply gradle.properties]: Added.

File: pravega-client-examples/README.md (hunk 3 above)
Comment (RaulGracia) [reply pravega-client-examples/README.md]: Actually I realized that the README version was not the correct one (which has the `streamcuts` sample execution instructions in the same order that the main table). Thanks.

File: scenarios/turbine-heat-processor/src/main/dist/bin/run-example (hunk 4 above)
Comment (RaulGracia) [reply scenarios/turbine-heat-processor/src/main/dist/bin/run-example]: Added.

File: scenarios/turbine-heat-sensor/src/main/java/io/pravega/turbineheatsensor/TurbineHeatSensor.java (hunk 5 above)
Comment (RaulGracia) [reply scenarios/turbine-heat-sensor/src/main/java/io/pravega/turbineheatsensor/TurbineHeatSensor.java]: Added.

File: README.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (Prabha-Veerubhotla) [reply README.md]: ok.

File: pravega-client-examples/README.md [hunk 14]
```diff
@@ -104,29 +85,52 @@ All args are optional, if not included, the defaults are:
  * scope - "examples"
  * name - "someStream"
  * uri - "tcp://127.0.0.1" (the URI to one of the controller nodes
+ 
+## `noop`
+ 
+ An example of a simple reader that continually reads the contents of any `Stream`. A binary serializer is used so it 
+ works against any event types. The sample emits basic information about number of events/bytes read every 30 seconds. 
+
```
Comment (Prabha-Veerubhotla) [inline pravega-client-examples/README.md]: add ###Execution here

File: pravega-client-examples/README.md [hunk 15]
```diff
@@ -104,29 +85,52 @@ All args are optional, if not included, the defaults are:
...
-Use the simple DSL to GET, PUT, REMOVE keys from the SharedConfig object identified by scope and name.
+Use the simple DSL to `GET`, `PUT`, `REMOVE` keys from the `SharedConfig` object identified by 
+scope and name. It is worthwhile to launch two or more separate CLIs in separate windows using 
+the same settings and observe how changes in one CLI process are not visible in another CLI 
+process until that other CLI process invokes `REFRESH`.
 
-It is worthwhile to launch two or more separate CLIs in separate windows using the same settings and observe how changes in one
-CLI process are not visible in another CLI process until that other CLI process invokes REFRESH.
+## `streamcuts`
+This application aims at demonstrating the use of `StreamCut`s four bounded processing
```
Comment (Prabha-Veerubhotla) [inline pravega-client-examples/README.md]: four or for?

File: pravega-client-examples/README.md [hunk 16]
```diff
@@ -104,29 +85,52 @@ All args are optional, if not included, the defaults are:
...
+on multiple `Stream`s. At the moment, the application contains two examples accessible via
+command line interface: i) Simple example: The user decides which `Stream` slices s/he wants 
+to read from all the st`Stream`reams by specifying indexes, and the application prints these slices
```
Comment (Prabha-Veerubhotla) [inline pravega-client-examples/README.md]: `Stream`

File: README.md [hunk 17]
```diff
@@ -11,56 +11,66 @@ developer guide](http://pravega.io).
...
 
-## Flink-connector Examples
+## Flink Connector Examples
 | Example Name  | Description  | Language |
 | ------------- |:-----| :-----|
-| `wordcount` | Counting the words continuously from a Pravega `Stream` to demonstrate the usage of Flink connector for Pravega. | [Java](https://github.com/pravega/pravega-samples/tree/master/flink-examples/src/main/java/io/pravega/examples/flink/wordcount)
+| `wordcount` | Counting the words continuously from a Pravega `Stream` to demonstrate the usage of Flink connector for Pravega. | [Java](https://github.com/pravega/pravega-samples/tree/master/flink-connector-examples/src/main/java/io/pravega/example/flink/wordcount)
 
-## Hadoop-connector Examples
+Please, find the related documentation and instructions [here](https://github.com/pravega/pravega-samples/tree/master/flink-connector-examples).
```
Comment (fpj) [inline README.md]: Rephrase it to: `The related documentation and instructions are [here]...`

File: README.md [hunk 18]
```diff
@@ -11,56 +11,66 @@ developer guide](http://pravega.io).
...
+| [`turbineheatprocessor`](https://github.com/pravega/pravega-samples/tree/master/scenarios/turbine-heat-processor) | A Flink streaming application for processing temperature data from a Pravega stream produced by the `turbineheatsensor` app. The application computes a daily summary of the temperature range observed on that day by each sensor. | [Java](https://github.com/pravega/pravega-samples/tree/master/scenarios/turbine-heat-processor/src/main/java/io/pravega/turbineheatprocessor), [Scala](https://github.com/pravega/pravega-samples/tree/master/scenarios/turbine-heat-processor/src/main/scala/io/pravega/turbineheatprocessor)
+| [`anomaly-detection`](https://github.com/pravega/pravega-samples/tree/master/scenarios/anomaly-detection) | A Flink streaming application for detecting anomalous input patterns using a finite-state machine. | [Java](https://github.com/pravega/pravega-samples/tree/master/scenarios/anomaly-detection/src/main/java/io/pravega/anomalydetection)
 
 
 # Build Instructions
 
+
+The `pravega-samples` project is prepared for working with [release artifacts](https://github.com/pravega/pravega/releases) of Pravega components,
```
Comment (fpj) [inline README.md]: These instructions are bit confusing because they are not super specific about how to configure the versions and so on. I suggest we start with instructions for how to run out of the box, and have some additional discussion with concrete examples to configure the samples to use a different pravega version.

File: README.md [hunk 19]
```diff
@@ -11,56 +11,66 @@ developer guide](http://pravega.io).
...
+to work with `master` snapshots artifacts published in our [JFrog repository](https://oss.jfrog.org/artifactory/jfrog-dependencies/io/pravega/).
+You can easily do this by specifying the desired artifact version (`pravegaVersion`, `flinkConnectorVersion`,
+`hadoopConnectorVersion`) in the `gradle.properties` file. Thus, if you look for a quick start 
+you may skip the _optional_ sections; otherwise, if you want to have fun building the different projects from source,
+please read the next sections. 
+
+
 ## Pre-requisites
 
 * Java 8+
```
Comment (fpj) [inline README.md]: I don't think we support Java 9, although we should by now.

File: README.md [hunk 20]
```diff
@@ -11,56 +11,66 @@ developer guide](http://pravega.io).
...
 
-For [release builds](https://github.com/pravega/pravega/releases) of Pravega, 
-the artifacts will already be in Maven Central and you will not need to run this step. 
-In this sense, you also have the option to download the Pravega `master` snapshots published 
-in our [JFrog repository](https://oss.jfrog.org/artifactory/jfrog-dependencies/io/pravega/).
-
-Conversely, if you want to build Pravega from source, you may need to generate the 
+If you want to build Pravega from source, you may need to generate the 
 latest Pravega `jar` files and install them to your local Maven repository. 
 To this end, please run the following commands:
```
Comment (fpj) [inline README.md]: Replace `To this end` with `To build Pravega from sources and it use it here`.

File: scenarios/turbine-heat-processor/gradle.properties [hunk 21]
```diff
@@ -0,0 +1,15 @@
+/*
```
Comment (fpj) [inline scenarios/turbine-heat-processor/gradle.properties]: I think this is the right formatting for this file, no?

```
#
# Copyright (c) 2017 Dell Inc., or its subsidiaries. All Rights Reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
```

File: scenarios/turbine-heat-processor/src/main/dist/bin/run-example [hunk 22]
```diff
@@ -0,0 +1,3 @@
+#!/usr/bin/env bash
```
Comment (fpj) [inline scenarios/turbine-heat-processor/src/main/dist/bin/run-example]: Missing license header.

File: scenarios/turbine-heat-sensor/gradle.properties [hunk 23]
```diff
@@ -0,0 +1,14 @@
+/*
```
Comment (fpj) [inline scenarios/turbine-heat-sensor/gradle.properties]: Same comment about the license header format for this file.

File: pravega-client-examples/README.md (hunk 14 above)
Comment (RaulGracia) [reply pravega-client-examples/README.md]: Added.

File: pravega-client-examples/README.md (hunk 15 above)
Comment (RaulGracia) [reply pravega-client-examples/README.md]: This was a typo, it should be `for`.

File: pravega-client-examples/README.md (hunk 16 above)
Comment (RaulGracia) [reply pravega-client-examples/README.md]: Fixed.

File: README.md (hunk 17 above)
Comment (RaulGracia) [reply README.md]: Done.

File: README.md (hunk 18 above)
Comment (RaulGracia) [reply README.md]: I tried to clarify the build instructions by distinguishing the `out-of-the-box` from the `build-from-source` options. All the mentions to `gradle.properties` are in the latter option. I also added an example of how to change the `gradle.properties` file to configure source builds. Let me know if the instructions are easier to follow now.

File: README.md (hunk 19 above)
Comment (RaulGracia) [reply README.md]: At the moment, I set `Java 8`.

File: README.md (hunk 20 above)
Comment (RaulGracia) [reply README.md]: Done.

File: scenarios/turbine-heat-processor/gradle.properties (hunk 21 above)
Comment (RaulGracia) [reply scenarios/turbine-heat-processor/gradle.properties]: Done.

File: scenarios/turbine-heat-processor/src/main/dist/bin/run-example (hunk 22 above)
Comment (RaulGracia) [reply scenarios/turbine-heat-processor/src/main/dist/bin/run-example]: Added.

File: scenarios/turbine-heat-sensor/gradle.properties (hunk 23 above)
Comment (RaulGracia) [reply scenarios/turbine-heat-sensor/gradle.properties]: Added.

**Comment by RaulGracia** [comment]
The changes of this PR have been moved to #105 to target `develop` branch.
